src/cargo/ops/registry/mod.rs

```diff
@@ -325,39 +325,52 @@ pub(crate) struct RegistrySourceIds {
 
 /// If this set of packages has an unambiguous publish registry, find it.
 pub(crate) fn infer_registry(pkgs: &[&Package]) -> CargoResult<Option<RegistryOrIndex>> {
-    if pkgs[1..].iter().all(|p| p.publish() == pkgs[0].publish()) {
-        // If all packages have the same publish settings, we take that as the default.
-        match pkgs[0].publish().as_deref() {
-            Some([unique_pkg_reg]) => {
-                Ok(Some(RegistryOrIndex::Registry(unique_pkg_reg.to_owned())))
-            }
-            None | Some([]) => Ok(None),
-            Some(regs) => {
-                let mut regs: Vec<_> = regs.iter().map(|s| format!("\"{}\"", s)).collect();
-                regs.sort();
-                regs.dedup();
-                // unwrap: the match block ensures that there's more than one reg.
-                let (last_reg, regs) = regs.split_last().unwrap();
-                bail!(
-                    "--registry is required to disambiguate between {} or {} registries",
-                    regs.join(", "),
-                    last_reg
-                )
+    // Ignore "publish = false" packages while inferring the registry.
+    let pkgs: Vec<_> = pkgs
+        .iter()
+        .filter(|p| p.publish() != &Some(Vec::new()))
+        .collect();
+
+    match &pkgs[..] {
+        [] => Ok(None),
+        [first, rest @ ..] => {
```

Comment:
If these are our only cases, this seems better served as a guard statement.  This causes a lot of unnecessary rightward drift and it gives the impression there will be other interesting cases when there aren't, breaking users expectations when reading the code, making it harder to read.

Reply:
By guard statement, I was referring to an early return, something like changing
```rust
if let Some((first, rest)) = publishable_pkgs.split_first() {
```
to
```rust
let Some((first, rest)) = publishable_pkgs.split_first()  else {
    return Ok(None);
};
```